governator-core/src/main/java/com/netflix/governator/internal/PreDestroyMonitor.java

```diff
@@ -0,0 +1,220 @@
+package com.netflix.governator.internal;
+
+import java.lang.annotation.Annotation;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.SoftReference;
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import com.google.inject.Binding;
+import com.google.inject.Key;
+import com.google.inject.Provider;
+import com.google.inject.Scope;
+import com.google.inject.spi.BindingScopingVisitor;
+import com.netflix.governator.LifecycleAction;
+import com.netflix.governator.ManagedInstanceAction;
+
+/**
+ * Monitors managed instances and invokes cleanup actions when they
+ * become unreferenced
+ * 
+ * @author tcellucci
+ *
+ */
+public class PreDestroyMonitor implements AutoCloseable {
+    
+     
+    private static class ScopeCleanupMarker {
+    }
+    
+    private final Deque<Callable<Void>> cleanupActions = new ConcurrentLinkedDeque<>();
+    
+    private final Map<Integer, ScopeCleanupAction> scopedCleanupActions = new LinkedHashMap<>();
+    private final Map<Class<? extends Annotation>, Scope> scopeBindings;
+    private final ReferenceQueue<ScopeCleanupMarker> markerReferenceQueue = new ReferenceQueue<>();
+    private final ExecutorService reqQueueExecutor = Executors.newSingleThreadExecutor();
+    
+    private final Provider<ScopeCleanupMarker> markerProvider = new Provider<ScopeCleanupMarker>() {
+        @Override
+        public ScopeCleanupMarker get() {
+            return new ScopeCleanupMarker();
+        }
+    };
+    
+    public PreDestroyMonitor(Map<Class<? extends Annotation>, Scope> scopeBindings) {
+        this.scopeBindings = new HashMap<>(scopeBindings);
+        reqQueueExecutor.submit(new Runnable() {
+            public void run() {
+                while (true) {
+                    try {
+                        Reference<? extends ScopeCleanupMarker> ref = markerReferenceQueue.remove(500);
+                        if (ref != null && ref instanceof ScopeCleanupAction) { 
+                            Integer markerKey = ((ScopeCleanupAction)ref).getId();
+                            ScopeCleanupAction cleanupAction;
+                            synchronized(scopedCleanupActions) {
+                                cleanupAction = scopedCleanupActions.remove(markerKey);
+                            }
+                            if (cleanupAction != null) {
+                                cleanupAction.call();
+                            }
+                        }
+                    } 
+                    catch (InterruptedException e) {
```

Comment:
made change to exit if interrupted, that allows Executor.shutdownNow() to function properly